semgrep-core/src/engine/Specialize_formula.ml

```diff
@@ -31,11 +40,22 @@ let selector_equal s1 s2 = s1.mvar = s2.mvar
 (* Converter *)
 (*****************************************************************************)
 
+(* return list of "positive" x list of Not x list of Conds *)
+let split_and :
+    R.formula list -> R.formula list * R.formula list * R.metavar_cond list =
+ fun xs ->
+  xs
+  |> Common.partition_either3 (fun e ->
+         match e with
+         | R.Not (_, f) -> Middle3 f
+         | R.Leaf (R.MetavarCond (_, c)) -> Right3 c
```

Comment:
I feel like we do `R.Leaf (R.MetavarCond (...))` and `R.Leaf (R.P (...))` a lot (at least, in this one file). Very not important but makes me think we would do better to have formula be
```
type formula =
  | P of xpattern * inside option
  | And of tok * formula list (* see Match_rules.split_and() *)
  | Or of tok * formula list
  | Not of tok * formula
  | MetavarCond of tok * metavar_cond
```

Reply:
It also feels weird to me to think of MetavarCond as a leaf because you certainly can have more patterns in it, and it doesn't stand alone.

Reply:
I agree too. I think they may have been grouped with the idea that some functions could handle "leafs" uniformly, but actually patterns and conditions are (almost?) always handled differently.

Reply:
Hmm, while I agree, if I would do this, I would want to do the same in `Rule.formula`. I leave this one for now.